synapse/api/ratelimiting.py

```diff
@@ -16,75 +17,157 @@
 from typing import Any, Optional, Tuple
 
 from synapse.api.errors import LimitExceededError
+from synapse.util import Clock
 
 
 class Ratelimiter(object):
     """
-    Ratelimit message sending by user.
+    Ratelimit actions marked by arbitrary keys.
+
+    Args:
+        clock: A homeserver clock, for retrieving the current time
+        rate_hz: The long term number of actions that can be performed in a second.
+        burst_count: How many actions that can be performed before being limited.
     """
 
-    def __init__(self):
-        self.message_counts = (
-            OrderedDict()
-        )  # type: OrderedDict[Any, Tuple[float, int, Optional[float]]]
+    def __init__(self, clock: Clock, rate_hz: float, burst_count: int):
```

Comment:
I debated two things here:

* Making `rate_hz` / `burst_count` optional so we don't have the case where we define them as 0 to override them each time.
* Having two rate limiters (with one being a sub-class of the other) for the two situations (per key limit vs. per rate limiter limit).

I think these are both just bikeshedding at this point though and it isn't worth poking at more. I just wanted to write them down!

Reply:
Neat ideas, will keep them in mind if we decide to do another refactor at some point. Thanks!